src/Features/Core/Portable/ConvertLinq/AbstractConvertLinqProvider.cs

```diff
@@ -0,0 +1,143 @@
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.ConvertLinq
+{
+    internal abstract class AbstractConvertLinqProvider : CodeRefactoringProvider
+    {
+        public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+            {
+                return;
+            }
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var semanticModel = await document.GetSemanticModelAsync().ConfigureAwait(false);
+
+            await CreateAnalyzer(semanticModel, context.CancellationToken)
+                .ComputeRefactoringsAsync(context).ConfigureAwait(false);
+        }
+
+        protected abstract IAnalyzer CreateAnalyzer(SemanticModel semanticModel, CancellationToken cancellationToken);
+
+        protected interface IAnalyzer
+        {
+            Task ComputeRefactoringsAsync(CodeRefactoringContext context);
+        }
+
+        protected abstract class AnalyzerBase<TRefactor, TSource, TDestination> : IAnalyzer
+            where TRefactor : SyntaxNode
+            where TSource : SyntaxNode
+            where TDestination : SyntaxNode
+        {
+            protected readonly SemanticModel _semanticModel;
+            protected readonly CancellationToken _cancellationToken;
+
+            public AnalyzerBase(SemanticModel semanticModel, CancellationToken cancellationToken)
+            {
+                _semanticModel = semanticModel;
+                _cancellationToken = cancellationToken;
+            }
+
+            public async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+            {
+                var document = context.Document;
+                var cancellationToken = context.CancellationToken;
+                var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                var refactor = FindNodeToRefactor(root, context);
+                if (refactor == null)
+                {
+                    return;
+                }
+
+                if (refactor.ContainsDiagnostics)
+                {
+                    return;
+                }
+
+                if (TryConvert(refactor, out var result))
+                {
+                    if (!result.IsValid())
+                    {
+                        return;
+                    }
+
+                    context.RegisterRefactoring(new MyCodeAction(Title, c => UpdateDocumentAsync(root, document, result)));
+                }
+            }
+
+            protected abstract bool TryConvert(TRefactor refactor, out DocumentUpdate documentUpdate);
+
+            protected virtual TRefactor FindNodeToRefactor(SyntaxNode root, CodeRefactoringContext context) =>
+                root.FindNode(context.Span).FirstAncestorOrSelf<TRefactor>();
```

Comment:
this seems *super* broad.  In general, in the IDE, we try to not do this so that the user isn't spammed with the lightbulb too much (for example, when they're deep inside a construct and could not care less about mutating it).

For queries, it probably makes sense to only offer this on hte very first token of the query. *or* potentially on the first token of any query clause.

Reply:
It should display the refactoring only for the whole QueryExpression or for its first child which are FromClause or the top QueryBody. So, maybe it is the opposite narrow not broad. What do you think?

---
In reply to: [175268080](https://github.com/dotnet/roslyn/pull/25362#discussion_r175268080) [](ancestors = 175268080)

Reply:
> It should display the refactoring only for the whole QueryExpression 

That is super massive.  A query could be many lines long and contain arbitrarily complex expressions.

We explicitly try to *not* support refactorings on the entire node of a construct that can be arbitrarily large.